src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/MapToRecordTransformer.java

```diff
@@ -21,72 +21,90 @@
     }
 
     Schema.Type fieldSchemaType = fieldSchema.getType();
 
     // empty string is considered null in csv
-    if (Strings.isNullOrEmpty(value)) {
+    if (value instanceof String && Strings.isNullOrEmpty((String) value)) {
       return null;
     }
 
     Schema.LogicalType logicalType = fieldSchema.getLogicalType();
     if (fieldSchema.getLogicalType() != null) {
-      return SalesforceTransformUtil.transformLogicalType(field.getName(), logicalType, value);
+      return SalesforceTransformUtil.transformLogicalType(fieldName, logicalType, String.valueOf(value));
     }
 
     switch (fieldSchemaType) {
       case NULL:
         return null;
       case BOOLEAN:
-        return Boolean.parseBoolean(value);
+        return Boolean.parseBoolean(castValue(value, fieldName, String.class));
       case INT:
-        return Integer.parseInt(value);
+        return Integer.parseInt(castValue(value, fieldName, String.class));
       case LONG:
-        return Long.parseLong(value);
+        return Long.parseLong(castValue(value, fieldName, String.class));
       case FLOAT:
-        return Float.parseFloat(value);
+        return Float.parseFloat(castValue(value, fieldName, String.class));
       case DOUBLE:
-        return Double.parseDouble(value);
+        return Double.parseDouble(castValue(value, fieldName, String.class));
       case STRING:
         return value;
+      case RECORD:
+        Map<String, String> recordValues = castGeneric(castValue(value, fieldName, Map.class));
+        StructuredRecord.Builder nestedBuilder = StructuredRecord.builder(fieldSchema);
+        Objects.requireNonNull(fieldSchema.getFields()).forEach(
+          field -> transformRecord(fieldSchema, recordValues, nestedBuilder));
+        return nestedBuilder.build();
+      case ARRAY:
+        List<Map<String, String>> list = castGeneric(castValue(value, fieldName, List.class));
+        Schema componentSchema = Objects.requireNonNull(fieldSchema.getComponentSchema());
+        return list.stream()
+          .map(map -> convertValue(fieldName, map, componentSchema))
+          .collect(Collectors.toList());
     }
 
     throw new UnexpectedFormatException(
-      String.format("Unsupported schema type: '%s' for field: '%s'. Supported types are 'boolean, int, long, float," +
-        "double and string'.", field.getSchema(), field.getName()));
+      String.format("Unsupported schema type: '%s' for field: '%s'. Supported types are 'boolean, int, long, float,"
+                      + "double, string, record, array'.", fieldSchema, fieldName));
+  }
+
+  private <T> T castValue(Object value, String fieldName, Class<T> clazz) {
+    if (clazz.isAssignableFrom(value.getClass())) {
+      return clazz.cast(value);
+    }
+    throw new UnexpectedFormatException(
+      String.format("Field '%s' is not of expected type '%s'", fieldName, clazz.getSimpleName()));
   }
 
+  @SuppressWarnings("unchecked")
+  private <T> T castGeneric(Object value) {
```

Comment:
this doesn't seem any different than just casting. Or is the purpose to avoid warnings?